### ISSUE TITLE
Capture request and response bodies

### ISSUE THREAD

**Issue by pavolloffay** [open-telemetry/semantic-conventions#857]
**What are you trying to achieve?**

I want to capture request and response bodies. The content can be used for business transaction monitoring and security.

To move this forward we need:
* data semantics e.g. `http.request.body`/`http.response.body` or be transport agnostic e.g. `request.body`, `response.body`.
* a configuration for auto instrumentations e.g. `OTEL_INSTRUMENTATION_CAPTURE_REQUEST_BODY=bool`

**Additional context.**

* business transaction monitoring http://trasier.com/
* capture request and response bodies in Java https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1317

**Comment by julealgon** [comment]
> Should the payload be a span event instead of a span attribute? Most backends have higher size limits for span events.

That doesn't make sense from a semantics point of view though... a piece of request/response data is not "an event" IMHO. And the HTTP span itself already represents a request or response, so there is no room to define a "send event" inside of that in my mind.

**Comment by chameleon82** [comment]
I would prefer to have body and other request/response attributes as a structured log. Moreover, to have the ability to separate them into different pipelines from other logs (different security and rotation policies). It is make sense to me to be able to select them by certain parameters and groups and able to process via anomaly detection tools. 
Currently, I do it this way already, but I had to write extra instruments and believe spec here will be helpful

**Comment by btsinfo** [comment]
If we consider observability from an enterprise solution perspective, the demand makes perfect sense and proves to be highly useful. However, if we take the perspective of agents alone, this demand might seem less relevant.

I assist companies in establishing enterprise-oriented observability. Managing performance, logs, and metrics is only the first step in an IT for IT approach. The real challenge lies in progressing to higher layers, namely an IT to Business approach.

At these higher levels, information derived from observability becomes strategically significant and even represents a considerable opportunity. By integrating use cases focused on real-time security with a business-oriented and customer behavior perspective, it becomes essential to have enriched data. This enables companies to capitalize on observability as a sub-datalake, integrated into the organization's global datalake.

In summary, it is crucial and timely to have this capability to transform observability into a strategic lever that serves the company.

**Comment by SylvainJuge** [comment]
Hi !

We (Elastic) are interested into moving forward the topic of capturing HTTP request bodies.

I have opened a PoC PR to implement this in the Java Instrumentation: https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/16006.
This currently captures request body first 1k bytes as a string in the `http.request.body.text` attribute.

I don't think this is the best place to discuss the properties and implications of such a feature, but we have to do it anyway just to ensure we are on the same page. We are already aware of the following:
- this must be an opt-in feature, this has obvious overhead costs
- sensitive data may be captured, so users must at least be aware of it
- content capture must be limited in size, this also implies that it may be truncated and thus harder/impossible to parse
- ability to sanitize or discard the captured values (for example to only capture it when span is an error or with a specific content type) can be implemented by selectively filtering/modifying attributes either on sdk or collector level (just like any other type of sanitization/redaction).

The feature implementation has implications on the semconv attribute name and expected content, we have identified a few possible options
- capture body as text, in this case the implementation must detect the data format and decode it properly. The main benefit of this approach is that the data is directly readable by end-users.
- capture body as raw bytes, using a base64 encoded string as recommended [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/README.md#byte-arrays), this simplifies the capture and potential truncation, adds some overhead, and leaves the decoding and conversion to the collector/backend, being able to do so would also likely require to also capture the `content-type` header of the request, maybe it would also make sense to automatically capture `http.request.header.content-type` when doing so.

From the perspective of implementation in the SDKs/instrumentation, I think the base64 encoded approach is probably the simplest and most straightforward as we don't have to deal with implementing duplicated heuristics to guess the content as text and we can leave that to a common implementation later in the data pipeline.

**Comment by trask** [comment]
> * capture body as raw bytes, using a base64 encoded string as recommended [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/README.md#byte-arrays)

This recommendation is under the section "AnyValue representation for non-OTLP protocols".

As of the most recent SDK release, we have true byte array attributes (no need for base64).

**Comment by trask** [comment]
> I don't think this is the best place to discuss the properties and implications of such a feature, but we have to do it anyway just to ensure we are on the same page.

It's probably a good time to open a semconv PR proposal since you have a prototype.

**Comment by NimrodAvni78** [comment]
@trask @SylvainJuge we also have [plans in OBI](https://github.com/open-telemetry/opentelemetry-ebpf-instrumentation/issues/396) to implement extracting http request and response bodies, so this will be useful for us.
I can support / open the proposal as well

**Comment by SylvainJuge** [comment]
I have just opened https://github.com/open-telemetry/semantic-conventions/pull/3521 with a proposal to capture it as `any` type and implement heuristics to capture it as string.

@NimrodAvni78 in your context I am not sure those "text heuristics" would be possible, but it could however still be captured as binary if needed.

**Comment by NimrodAvni78** [comment]
@SylvainJuge we also are planning to use the content-type header to detect if we can capture the http payload, since it make the most sense to capture only textual payloads at the moment

we are also planning on providing ways to obfuscate sensitive data based on the content type, for example in json payloads you want to remove a value of a specific field based on a jsonPath for example, I don't know if it should be part of the spec or just an implementation detail of instrumentations

We are also planning to add the response in the same step as we do the request, so i would also vouch for adding the http response payload to the suggestion

**Comment by SylvainJuge** [comment]
@NimrodAvni78 I've updated https://github.com/open-telemetry/semantic-conventions/pull/3521 to also cover request and response.

For dealing with obfuscation of sensitive data, I don't know if there is a dedicated spec for that somewhere, but in this particular case we will need to support potentially partial data, for example if a json is truncated a simple json parser would likely not work as expected. I know dealing with this is best from security when dealt at the source, however having an implementation in the collector would allow to use it more broadly without having to re-implement it everywhere.
